### PR TITLE
Support proxying WebSocket connections.

### PR DESCRIPTION
**Pull request by chirino**

### CODE REVIEW

File: oauthproxy.go [hunk 1]
```diff
@@ -92,7 +94,12 @@ func (u *UpstreamProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		r.Header.Set("GAP-Auth", w.Header().Get("GAP-Auth"))
 		u.auth.SignRequest(r)
 	}
-	u.handler.ServeHTTP(w, r)
+	if u.wsHandler != nil && r.Header.Get("Upgrade") == "websocket" {
```
Comment (mrogers950) [inline oauthproxy.go]: We might want to check for the Connection header being set to "upgrade" as well.

File: oauthproxy.go [hunk 2]
```diff
@@ -146,6 +153,23 @@ func NewFileServer(path string, filesystemPath string) (proxy http.Handler) {
...
+	if err != nil {
+		log.Fatal("Failed to initialize Reverse Proxy: ", err)
+	}
+	if !opts.PassHostHeader {
+		setProxyUpstreamHostHeader(proxy, u)
+	} else {
+		setProxyDirector(proxy)
+	}
+
+	wsURL := &url.URL{Scheme: "ws://", Host: u.Host}
```
Comment (mrogers950) [inline oauthproxy.go]: We should support wss also but that support can follow this PR.

File: oauthproxy.go [hunk 3]
```diff
@@ -146,6 +153,23 @@ func NewFileServer(path string, filesystemPath string) (proxy http.Handler) {
...
+	wsProxy := wsutil.NewSingleHostReverseProxy(wsURL)
```
Comment (mrogers950) [inline oauthproxy.go]: Do we always want to set up the ws handler or should we make this configurable?

Comment (chirino) [reply oauthproxy.go]: Added an option for that.  But defaults to enabled.

File: oauthproxy.go (hunk 2 above)
Comment (chirino) [reply oauthproxy.go]: Added some initial support for it.  Should use ws or wss based on if it's an http or https upstream.

File: oauthproxy.go (hunk 1 above)
Comment (chirino) [reply oauthproxy.go]: Good idea.

**Comment by chirino** [comment]
@mrogers950 do you need anything else to make this PR acceptable?

**Comment by mrogers950** [comment]
@chirino looks good, though it would be nice to have wss included in the unit tests as well.

File: options.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ import (
 // Configuration Options that can be set by Command Line Flag, or Config File
 type Options struct {
 	ProxyPrefix      string        `flag:"proxy-prefix" cfg:"proxy-prefix"`
+	ProxyWebSockets  bool          `flag:"proxy-websockets" cfg:"proxy_websockets"`
```
Comment (mrogers950) [inline options.go]: You'll also need to add to the flagset in main.go for this option

File: oauthproxy.go [hunk 5]
```diff
@@ -165,8 +165,13 @@ func NewWebSocketOrRestReverseProxy(u *url.URL, opts *Options, auth hmacauth.Hma
...
 	}
 
-	wsURL := &url.URL{Scheme: "ws://", Host: u.Host}
-	wsProxy := wsutil.NewSingleHostReverseProxy(wsURL)
+	// this should give us a wss:// scheme if the url is https:// based.
+	var wsProxy *wsutil.ReverseProxy = nil
+	if opts.ProxyWebSockets {
+		wsScheme := "ws" + strings.TrimPrefix(u.Scheme, "http")
+		wsURL := &url.URL{Scheme: wsScheme, Host: u.Host}
+		wsProxy = wsutil.NewSingleHostReverseProxy(wsURL)
```
Comment (mrogers950) [inline oauthproxy.go]: I think you will also need to set up wsProxy.TLSClientConfig with the upstream CAs if using wss, e.g.,

		if wsScheme == "wss" && len(opts.UpstreamCAs) > 0 {
			pool, err := util.GetCertPool(opts.UpstreamCAs, false)
			if err != nil {
				log.Fatal("Failed to fetch CertPool: ", err)
			}
			wsProxy.TLSClientConfig = &tls.Config{
				RootCAs: pool,
			}
		}

**Comment by chirino** [comment]
@mrogers950 if you have an example of a https reverse proxy test, I'd be happy to adapt it to also test the wss proxying.

**Comment by mrogers950** [comment]
@chirino I'm working on end-to-end testing for oauth-proxy with TLS tests, so we can add wss tests later.
/lgtm
@simo5 any feedback?

**Comment by simo5** [comment]
@chirino can you squash all the commits *except* for the vendoring stuff (leave that as a separate commit).

Then it's good to go in.

**Comment by chirino** [comment]
@simo5 squashd.

**Comment by chirino** [comment]
@simo5 @mrogers950 you guys need anything else?

**Comment by simo5** [comment]
@chirino not it looks good, I'll let @mrogers950 decide whether to push now or wait until we have CI and can run tests.

**Comment by mrogers950** [comment]
Looks good to me, thanks for your contribution @chirino !

**Comment by chirino** [comment]
@mrogers950  Many thanks.  How often do you guys push docker image releases?

**Pull request merged**